tiflash crash frequently with error of TryFlushData faild

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
[2022/08/11 01:56:46.760 +08:00] [ERROR] [Exception.cpp:85] ["uint8_t DB::TryFlushData(DB::EngineStoreServerWrap *, uint64_t, uint8_t, uint64_t, uint64_t):Code: 0, e.displayText() = DB::Exception: region not found when trying flush, e.what() = DB::Exception, Stack trace:

       0x1a91a98    DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+27859608]
                    dbms/src/Common/Exception.h:41
       0x651ab64    DB::KVStore::canFlushRegionDataImpl(std::__1::shared_ptr<DB::Region> const&, unsigned char, bool, DB::TMTContext&, DB::RegionTaskLock const&, unsigned long, unsigned long) [tiflash+106015588]
                    dbms/src/Storages/Transaction/KVStore.cpp:348
       0x651ad18    DB::KVStore::tryFlushRegionData(unsigned long, bool, DB::TMTContext&, unsigned long, unsigned long) [tiflash+106016024]
                    dbms/src/Storages/Transaction/KVStore.cpp:341
       0x65310f8    TryFlushData [tiflash+106107128]
                    dbms/src/Storages/Transaction/ProxyFFI.cpp:150
  0xffffacadfafc    _$LT$raftstore..engine_store_ffi..observer..TiFlashObserver$u20$as$u20$raftstore..coprocessor..AdminObserver$GT$::pre_exec_admin::hfd267e9df0a1ada2 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+30669564]
  0xffffac476d38    raftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::ha0396bd1778fd5cd [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+23948600]
  0xffffac475f34    raftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::h0aac2d505272f3fb [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+23945012]
  0xffffac491034    raftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h74c0b72499b90339 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+24055860]
  0xffffac4939f8    raftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::hc57505dc8bb700b6 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+24066552]
  0xffffac496324    _$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::ha4565cf6507a3216 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+24077092]
  0xffffac0a5508    batch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::hc14ea824bead376d [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+19944712]
  0xffffac77136c    std::sys_common::backtrace::__rust_begin_short_backtrace::h80953f8e29109df0 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+27071340]
  0xffffabf96908    core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h8b14e8f2aab21868 [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+18835720]
  0xfffface2f01c    std::sys::unix::thread::Thread::new::thread_start::h4e461bcfb14a5bff [[libtiflash_proxy.so](http://libtiflash_proxy.so/)+34140188]
  0xffffaac37d38    start_thread [libpthread.so.0+32056]
  0xffffaab3f690    thread_start [libc.so.6+915088]"] [thread_id=93]
[2022/08/11 01:56:46.762 +08:00] [ERROR] [BaseDaemon.cpp:377] [BaseDaemon:########################################] [thread_id=239]
[2022/08/11 01:56:46.763 +08:00] [ERROR] [BaseDaemon.cpp:378] ["BaseDaemon:(from thread 178) Received signal Segmentation fault(11)."] [thread_id=239]
[2022/08/11 01:56:46.763 +08:00] [ERROR] [BaseDaemon.cpp:408] ["BaseDaemon:Address: 0x1390"] [thread_id=239]
[2022/08/11 01:56:46.763 +08:00] [ERROR] [BaseDaemon.cpp:423] ["BaseDaemon:Address not mapped to object."] [thread_id=239]
[2022/08/11 01:56:46.765 +08:00] [WARN] [CoprocessorHandler.cpp:143] ["CoprocessorHandler:RegionException: region 22449, message: NOT_FOUND"] [thread_id=117]
[2022/08/11 01:56:46.765 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:
       0x5a1cad0    faultSignalHandler(int, siginfo_t*, void*) [tiflash+94489296]
                    libs/libdaemon/src/BaseDaemon.cpp:221
  0xffffae11066c    <unknown symbol> [+1644]"] [thread_id=239]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)

--- a/dbms/src/Storages/Transaction/KVStore.cpp
+++ b/dbms/src/Storages/Transaction/KVStore.cpp
@@ -340,6 +340,10 @@ bool KVStore::tryFlushRegionData(UInt64 region_id, bool try_until_succeed, TMTCo
     const RegionPtr curr_region_ptr = getRegion(region_id);
     if (curr_region_ptr == nullptr)
     {
+        /// If we can't find region here, we return true so proxy can trigger a CompactLog.
+        /// The triggered CompactLog will be handled by `handleUselessAdminRaftCmd`,
+        /// and result in a `EngineStoreApplyRes::NotFound`.
+        /// Proxy will print `region not found in engine-store, maybe have exec `RemoveNode` first`.
         LOG_FMT_WARNING(log, "region {} [index: {}, term {}], not exist when flushing, maybe have exec `RemoveNode` first", region_id, index, term);
         return true;
     }

--- a/dbms/src/Storages/Transaction/KVStore.cpp
+++ b/dbms/src/Storages/Transaction/KVStore.cpp
@@ -355,7 +355,7 @@ bool KVStore::canFlushRegionDataImpl(const RegionPtr & curr_region_ptr, UInt8 fl
 {
     if (curr_region_ptr == nullptr)
     {
-        throw Exception(fmt::format("region not found when trying flush", ErrorCodes::LOGICAL_ERROR));
+        throw Exception("region not found when trying flush", ErrorCodes::LOGICAL_ERROR);
     }
     auto & curr_region = *curr_region_ptr;
 

--- a/dbms/src/Storages/Transaction/KVStore.cpp
+++ b/dbms/src/Storages/Transaction/KVStore.cpp
@@ -344,7 +344,7 @@ bool KVStore::tryFlushRegionData(UInt64 region_id, bool try_until_succeed, TMTCo
         /// If we can't find region here, we return true so proxy can trigger a CompactLog.
         /// The triggered CompactLog will be handled by `handleUselessAdminRaftCmd`,
         /// and result in a `EngineStoreApplyRes::NotFound`.
-        /// Proxy will print `region not found in engine-store, maybe have exec `RemoveNode` first`.
+        /// Proxy will print this message and continue: `region not found in engine-store, maybe have exec `RemoveNode` first`.
         LOG_FMT_WARNING(log, "region {} [index: {}, term {}], not exist when flushing, maybe have exec `RemoveNode` first", region_id, index, term);
         return true;
     }

--- a/dbms/src/Storages/Transaction/KVStore.cpp
+++ b/dbms/src/Storages/Transaction/KVStore.cpp
@@ -338,6 +338,11 @@ bool KVStore::tryFlushRegionData(UInt64 region_id, bool try_until_succeed, TMTCo
 {
     auto region_task_lock = region_manager.genRegionTaskLock(region_id);
     const RegionPtr curr_region_ptr = getRegion(region_id);
+    if (curr_region_ptr == nullptr)
+    {
+        LOG_FMT_WARNING(log, "region {} [index: {}, term {}], not exist when flushing, maybe have exec `RemoveNode` first", region_id, index, term);
+        return true;
+    }
     return canFlushRegionDataImpl(curr_region_ptr, true, try_until_succeed, tmt, region_task_lock, index, term);
 }
 

--- a/dbms/src/Storages/Transaction/KVStore.cpp
+++ b/dbms/src/Storages/Transaction/KVStore.cpp
@@ -331,6 +331,7 @@ bool KVStore::needFlushRegionData(UInt64 region_id, TMTContext & tmt)
 {
     auto region_task_lock = region_manager.genRegionTaskLock(region_id);
     const RegionPtr curr_region_ptr = getRegion(region_id);
+    // TODO Should handle when curr_region_ptr is null.
     return canFlushRegionDataImpl(curr_region_ptr, false, false, tmt, region_task_lock, 0, 0);
 }
 

--- a/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
+++ b/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
@@ -1251,7 +1251,6 @@ void RegionKVStoreTest::testKVStore()
         request.mutable_compact_log();
         request.set_cmd_type(::raft_cmdpb::AdminCmdType::CompactLog);
         ASSERT_EQ(kvs.handleAdminRaftCmd(raft_cmdpb::AdminRequest{request}, std::move(response), 1999, 22, 6, ctx.getTMTContext()), EngineStoreApplyRes::NotFound);
-
     }
 }
 

--- a/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
+++ b/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
@@ -1244,6 +1244,14 @@ void RegionKVStoreTest::testKVStore()
         ASSERT_EQ(kvs.tryFlushRegionData(19, true, ctx.getTMTContext(), 0, 0), true);
         // Non existing region.
         ASSERT_EQ(kvs.tryFlushRegionData(1999, true, ctx.getTMTContext(), 0, 0), true);
+
+        raft_cmdpb::AdminRequest request;
+        raft_cmdpb::AdminResponse response;
+
+        request.mutable_compact_log();
+        request.set_cmd_type(::raft_cmdpb::AdminCmdType::CompactLog);
+        ASSERT_EQ(kvs.handleAdminRaftCmd(raft_cmdpb::AdminRequest{request}, std::move(response), 1999, 22, 6, ctx.getTMTContext()), EngineStoreApplyRes::NotFound);
+
     }
 }
 

--- a/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
+++ b/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
@@ -1242,6 +1242,8 @@ void RegionKVStoreTest::testKVStore()
         ASSERT_EQ(kvs.needFlushRegionData(19, ctx.getTMTContext()), true);
         // Force flush until succeed only for testing.
         ASSERT_EQ(kvs.tryFlushRegionData(19, true, ctx.getTMTContext(), 0, 0), true);
+        // Non existing region.
+        ASSERT_EQ(kvs.tryFlushRegionData(1999, true, ctx.getTMTContext(), 0, 0), true);
     }
 }
 

--- a/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
+++ b/dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp
@@ -1243,11 +1243,10 @@ void RegionKVStoreTest::testKVStore()
         // Force flush until succeed only for testing.
         ASSERT_EQ(kvs.tryFlushRegionData(19, true, ctx.getTMTContext(), 0, 0), true);
         // Non existing region.
+        // Flush and CompactLog will not panic.
         ASSERT_EQ(kvs.tryFlushRegionData(1999, true, ctx.getTMTContext(), 0, 0), true);
-
         raft_cmdpb::AdminRequest request;
         raft_cmdpb::AdminResponse response;
-
         request.mutable_compact_log();
         request.set_cmd_type(::raft_cmdpb::AdminCmdType::CompactLog);
         ASSERT_EQ(kvs.handleAdminRaftCmd(raft_cmdpb::AdminRequest{request}, std::move(response), 1999, 22, 6, ctx.getTMTContext()), EngineStoreApplyRes::NotFound);

